Sdem/further cleanup and simplification

GuillermoCasas opened the pull request:
Many changes to the json and some simplifications to the analysis_stage of SDEM. All tests run OK, including the PFEM one. I prefer to do this adaptation half way through the process, so you all can adat to it as soon as possible.

GuillermoCasas committed:
Simplifying interface + starting to group back-coupling-related params

GuillermoCasas committed:
Merge branch 'master' into SDEM/further-cleanup-and-simplification

GuillermoCasas committed:
Many simplifications. 

Removing variables and options

GuillermoCasas committed:
Tidying and simplificating

The constructor of simming_DEM_analysis has been particularly simplified. Also the json has been given more structure.

GuillermoCasas committed:
More structure to json

and more simplifications

philbucher commented on applications/SwimmingDEMApplication/custom_python/add_custom_utilities_to_python.cpp:
```diff
@@ -396,7 +396,7 @@ void  AddCustomUtilitiesToPython(pybind11::module& m){
 
     py::class_<BassetForceTools> (m, "BassetForceTools")
-        .def(py::init<>())
+        .def(py::init<Parameters&>())
```
`Kratos::Parameters` should be passed by-value according to riccardo due to how internally the storage of the json is handled

GuillermoCasas replied:
This will be addressed in its own PR

philbucher commented on applications/SwimmingDEMApplication/python_scripts/swimming_DEM_analysis.py:
```diff
@@ -33,12 +33,12 @@ def Say(*args):
     from KratosMultiphysics.mpi import *
```
this is deprecated
use the MPI-Core / dsta-communnicator instead, see the [tests](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/mpi/tests/test_mpi_data_communicator_python.py)

philbucher commented on applications/SwimmingDEMApplication/python_scripts/swimming_DEM_analysis.py:
```diff
@@ -33,12 +33,12 @@ def Say(*args):
 
     # DEM Application MPI
-    import DEM_procedures_mpi as DEM_procedures
+    import DEM_procedures_mpi as DP
```
this way of importing python-scripts will not be supported forever (onyl as long as the python-scripts in the apps are added to the python-path), better use the full-path

```suggestion
    import KratosMultiphysics.DEMApplication.DEM_procedures_mpi as DP
```

more details [see here](https://github.com/KratosMultiphysics/Kratos/wiki/Applications-as-python-modules)

GuillermoCasas replied:
Also to be done all at once in a future PR. There are many instances of this both in SDEM and DEM apps.

philbucher commented on applications/SwimmingDEMApplication/python_scripts/swimming_DEM_analysis.py:
```diff
@@ -82,85 +82,54 @@ def __init__(self, model, parameters = Parameters("{}")):
 
         # First, read the parameters generated from the interface
-        self.ReadDispersePhaseAndCouplingParameters()
+        import swimming_dem_default_input_parameters as only_swimming_defaults
```
```suggestion
        import KratosMultiphysics.SwimmingDEMApplication.swimming_dem_default_input_parameters as only_swimming_defaults
```

philbucher commented on applications/SwimmingDEMApplication/python_scripts/swimming_DEM_analysis.py:
```diff
@@ -788,12 +735,26 @@ def AssignKinematicViscosityFromDynamicViscosity(self):
 
+    def _GetDEMAnalysis(self):
+        if not hasattr(self, '_disperse_phase_analysis'):
+            import fluid_coupled_DEM_analysis as DEM_analysis
```
idem

philbucher commented on applications/SwimmingDEMApplication/python_scripts/swimming_DEM_analysis.py:
```diff
@@ -788,12 +735,26 @@ def AssignKinematicViscosityFromDynamicViscosity(self):
+
+    def _GetFluidAnalysis(self):
+        if not hasattr(self, '_fluid_phase_analysis'):
+            import DEM_coupled_fluid_dynamics_analysis as fluid_analysis
```
idem

philbucher commented on applications/SwimmingDEMApplication/python_scripts/swimming_DEM_procedures.py:
```diff
@@ -4,7 +4,7 @@
 import KratosMultiphysics.DEMApplication as DEMApp
 import KratosMultiphysics.SwimmingDEMApplication as SDEMApp
-import DEM_procedures
+import DEM_procedures as DP
```
same

philbucher commented on applications/SwimmingDEMApplication/python_scripts/swimming_DEM_solver.py:
```diff
@@ -1,24 +1,25 @@
 # Importing the Kratos Library
-import KratosMultiphysics
+from KratosMultiphysics import *
 from python_solver import PythonSolver
```
```suggestion
from KratosMultiphysics.python_solver import PythonSolver
```

philbucher commented on applications/SwimmingDEMApplication/python_scripts/swimming_DEM_solver.py:
```diff
@@ -1,24 +1,25 @@
-    KratosMultiphysics.Logger.PrintInfo("SwimmingDEM", *args)
-    KratosMultiphysics.Logger.Flush()
+    Logger.PrintInfo("SwimmingDEM", *args)
+    Logger.Flush()
```
flushing by default might slow the simulation down in MPI, depending on the cluster

philbucher commented on applications/SwimmingDEMApplication/python_scripts/swimming_dem_default_input_parameters.py:
```diff
@@ -5,129 +5,172 @@ def GetDefaultInputParameters():
+                "time_step"           : 0.001
+            },
+
+            "problem_data"     : {
```
+1 for consistency with other apps

philbucher reviewed (commented):
I made some comments for improvements
regarding the selection of MPI/OMP I recommend to use the `parallel_type` in the settings (this is done in Fluid, structure & others)
your solution seems to work with OpenMPI only, but many clusters have Intel-compilers, not sure if it would work there

GuillermoCasas committed:
Removing MPI imports

MPI is not currently working in SDEM. I am thus removing the deprecated calls altogether. When MPI is actually implemented I will do it properly.

maceligueta commented:
Let's dedicate an issue and PR for @philbucher's comments. I will open the issue as a reminder.
